src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs

```diff
@@ -20,15 +20,17 @@ public static class TextLoaderSaverCatalog
         /// <param name="hasHeader">Whether the file has a header.</param>
         /// <param name="allowSparse">Whether the file can contain numerical vectors in sparse format.</param>
         /// <param name="allowQuoting">Whether the file can contain column defined by a quoted string.</param>
+        /// <param name="trimWhitespace">Remove trailing whitespace from lines</param>
         /// <param name="dataSample">The optional location of a data sample. The sample can be used to infer column names and number of slots in each column.</param>
         public static TextLoader CreateTextLoader(this DataOperationsCatalog catalog,
             TextLoader.Column[] columns,
             char separatorChar = TextLoader.Defaults.Separator,
             bool hasHeader = TextLoader.Defaults.HasHeader,
             bool allowSparse = TextLoader.Defaults.AllowSparse,
             bool allowQuoting = TextLoader.Defaults.AllowQuoting,
+            bool trimWhitespace = TextLoader.Defaults.TrimWhitespace,
             IMultiStreamSource dataSample = null)
```

Comment:
The data sample is ***really important***. (See #2663.) If there are feature names, if there is a header, that's where they're coming from. It's really important. Burying it... what... *eight* levels deep seems a bit much. To bury it beyond such things as allowing sparsity and quoting itself seems really tough.

I might say the following order is most sensible:

1. Columns,
2. Separator char,
3. Has header,
4. Data sample,
5. Quoting,
6. Trimming,
7. Sparsity.

I might even think that 5, 6, and 7 could be relegated to advanced options, but no need to insist.

Reply:
Just got them reordered. Thanks for sharing the list!

---
In reply to: [259963185](https://github.com/dotnet/machinelearning/pull/2710#discussion_r259963185) [](ancestors = 259963185)

Reply:
I think that if hasHeader is true, then the dataSample should not be allowed to be null (because if it is, there can be no slot names). Does it make sense to check this? Another alternative is instead of having a bool argument, to have an `IMultiStreamSource header` instead. But on the other hand, if there are two `IMultiStreamSource`s that might be confusing. Not sure what the right signature is here. @ivanidzo4ka, @TomFinley, what do you think?

---
In reply to: [260017117](https://github.com/dotnet/machinelearning/pull/2710#discussion_r260017117) [](ancestors = 260017117,259963185)

Reply:
> I think that if hasHeader is true, then the dataSample should not be allowed to be null (because if it is, there can be no slot names). Does it make sense to check this? Another alternative is instead of having a bool argument, to have an `IMultiStreamSource header` instead. But on the other hand, if there are two `IMultiStreamSource`s that might be confusing. Not sure what the right signature is here. @Ivanidzo4ka, @TomFinley, what do you think?
> 
> In reply to: [260017117](https://github.com/dotnet/machinelearning/pull/2710#discussion_r260017117) [](ancestors = 260017117,259963185)

So, let's imagine that I declare that there is a header, but I provide no sample. I think this is fine. I am merely describing *how* to read data, and I am saying that when you read the data with this thing, that you will have a header to contend with. If headers were always only used for defining slot names, I might agree, but sometimes files have headers just because they have headers, and there's no larger reason for it.

The presence of absence of a sample does not mean that we cannot fully define the schema, even if header is true. (This is in contrast to some other options, like, "this is fixed size, but I would like you to auto-determine what that size is, which obviously requires a sample.)

So no check would be my reaction.